[SymmMem] Find NVSHMEM from system installation

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157513

Previously we only search for NVSHMEM from pip install location. 
This PR adds search in system locations deemed default by CMake.
Related: #157453 untars NVSHMEM into `/usr/local` on our CI machines.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @Skylion007

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu:
```diff
@@ -6,9 +6,18 @@
-#include <cuda_awbarrier_primitives.h>
 // Use torch's cub wrapper instead of CUDA's <cub/cub.cuh>, see #55292
 #include <ATen/cuda/cub.cuh>
+
+// NVSHMEM minimum SM arch
+#define _NVSHMEM_MIN_SM_ARCH 700
+
+// Some NVSHMEM device APIs do not compile on older SM archs
+#if defined(__CUDA_ARCH__) && (__CUDA_ARCH__ < _NVSHMEM_MIN_SM_ARCH)
+// Only include host APIs. See nvshmem.h for details.
+#define NVSHMEM_HOSTLIB_ONLY
+#endif  // Must be done before nvshmem.h is included
+
```
@atalman Solving the sm_52 compile issue.

Skylion007 commented on tools/setup_helpers/cmake.py:
```diff
@@ -346,9 +346,10 @@ def generate(
-            build_options["NVSHMEM_HOME"] = nvshmem_home
+        nvshmem_py_dir = py_lib_path + "/nvidia/nvshmem"
+        if os.path.exists(nvshmem_py_dir):
+            print("NVSHMEM wheel install found: " + nvshmem_py_dir)
```
This print should be removed before merge

kwen2501 committed:
Update on "[SymmMem] Find NVSHMEM from system installation"


Previously we only search for NVSHMEM from pip install location. 
This PR adds search in system locations deemed default by CMake.
Related: #157453 untars NVSHMEM into `/usr/local` on our CI machines.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k Skylion007 

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge -f "CI passed; removing a print statement"

Skylion007 commented:
I never actually exported NVSHMEM for the builder so this patch is missing one more part:
According to the logs:
```
-- NVSHMEM_HOME set to:  ''
-- NVSHMEM wheel installed at:  ''
-- NVSHMEM not found, not building with NVSHMEM support.
``` 

So we still need one more component sadly to properly set NVSHMEM...

Skylion007 reviewed (commented):
We still need some more NVSHMEM locations set

kwen2501 commented on caffe2/CMakeLists.txt:
```diff
@@ -989,23 +989,30 @@ elseif(USE_CUDA)
+    find_path(NVSHMEM_LIB_DIR
+      # In pip install case, the lib suffix is `.so.3` instead of `.so`
+      NAMES libnvshmem_host.so libnvshmem_host.so.3
+      PATHS $ENV{NVSHMEM_HOME}/lib ${NVSHMEM_PY_DIR}/lib
```
arg, I thought `find_path` would search for default paths if the file is not found from the specified paths.

kwen2501 commented:
@pytorchbot cherry-pick --onto release/2.8 -c Compilation

kwen2501 commented:
@pytorchbot cherry-pick --onto release/2.8 -c critical
